x-pack/plugins/security_solution/public/management/components/policies_selector/policies_selector.tsx

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { memo, useCallback, useMemo, useState, useEffect, ChangeEvent } from 'react';
+
+import { i18n } from '@kbn/i18n';
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiFilterGroup,
+  EuiPopover,
+  EuiPopoverTitle,
+  EuiFieldSearch,
+  EuiFilterButton,
+  EuiFilterSelectItem,
+  FilterChecked,
+  EuiText,
+} from '@elastic/eui';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { ImmutableArray, PolicyData } from '../../../../common/endpoint/types';
+
+export interface PoliciesSelectorProps {
+  policies: ImmutableArray<PolicyData>;
+  defaultIncludedPolicies?: string;
+  defaultExcludedPolicies?: string;
+  onChangeSelection: (items: Item[]) => void;
+}
+
+export interface Item {
+  name: string;
+  id?: string;
+  checked?: FilterChecked;
+}
+
+interface DefaultPoliciesByKey {
+  [key: string]: boolean;
+}
+
+const GLOBAL_ENTRIES = i18n.translate(
+  'xpack.securitySolution.management.policiesSelector.globalEntries',
+  {
+    defaultMessage: 'Global entries',
+  }
+);
+const UNASSIGNED_ENTRIES = i18n.translate(
+  'xpack.securitySolution.management.policiesSelector.unassignedEntries',
+  {
+    defaultMessage: 'Unassigned entries',
+  }
+);
+
+export const PoliciesSelector = memo<PoliciesSelectorProps>(
+  ({ policies, onChangeSelection, defaultExcludedPolicies, defaultIncludedPolicies }) => {
+    const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+    const [query, setQuery] = useState<string>('');
+    const [itemsList, setItemsList] = useState<Item[]>([]);
+
+    useEffect(() => {
+      const defaultIncludedPoliciesByKey: DefaultPoliciesByKey = defaultIncludedPolicies
+        ? defaultIncludedPolicies.split(',').reduce((acc, val) => ({ ...acc, [val]: true }), {})
+        : {};
+
+      const defaultExcludedPoliciesByKey: DefaultPoliciesByKey = defaultExcludedPolicies
+        ? defaultExcludedPolicies.split(',').reduce((acc, val) => ({ ...acc, [val]: true }), {})
+        : {};
+
+      const getCheckedValue = (id: string): FilterChecked | undefined =>
+        defaultIncludedPoliciesByKey[id]
+          ? 'on'
+          : defaultExcludedPoliciesByKey[id]
+          ? 'off'
+          : undefined;
+
+      setItemsList([
+        ...policies.map((policy) => ({
+          name: policy.name,
+          id: policy.id,
+          checked: getCheckedValue(policy.id),
+        })),
+        { name: GLOBAL_ENTRIES, id: 'global', checked: getCheckedValue('global') },
+        { name: UNASSIGNED_ENTRIES, id: 'unassigned', checked: getCheckedValue('unassigned') },
+      ]);
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [policies]);
+
+    const onButtonClick = () => {
```

Comment:
can you wrap this in `useCallback()`? Same for the `closePopover` below as well.

Reply:
Yeah, I thought that but following react docs: https://reactjs.org/docs/hooks-reference.html#usestate
It says:
```
React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.
```

As this functions only has setState actions I think is one of those cases you don't need to wrap it in a useCallback as it wouldn't be better in terms on performance.

Note: I'm using the setState param function to avoid memoize the current value. 
```
setIsPopoverOpen((prevIsPopoverOpen) => !prevIsPopoverOpen);
```

Reply:
🤔 - true. the setter function provided back by `useState()` is guaranteed to now change and thus you can omit that setter function from the dependencies of useCallback.

One of the benefits of using `useCallback()` is that the returned function is always the same until the dependencies change. Because `onButtonClick` is given as a prop to another component, and because its value (the function) is a new function instance (because its not wrapped in `useCallback()`), that will trigger a re-render of the downstream component (prop changed).

Reply:
☝️  ( 😁 ) - still optional

Reply:
Oh right, I'm still wondering if it will improve something as there is no changing state/prop related? So the returned function still being the same right?

Reply:
No, every time this component is rendered, the `onButtonClick` will be a new instance (function) because it's not cached. So when that is used below as a prop value, it will cause that comment below to also re-render (because the prop value is new (`oldOnButtonClick !== onButtonClick`)

Reply:
Argh! I see, got it. Thanks!!

Reply:
Done in: https://github.com/elastic/kibana/pull/106710/commits/6e92a6b1fbb3b4abb599b01e3e68b0892dadf624